justfile

```diff
@@ -1,55 +1,100 @@
+# just has no idiom for setting an default value for an environment variable
+# so we shell out, as we need VIRTUAL_ENV in the justfile environment
+export VIRTUAL_ENV  := `echo ${VIRTUAL_ENV:-.venv}`
+
+# TODO: make it /scripts on windows?
+export BIN := VIRTUAL_ENV + "/bin"
+export PIP := BIN + "/python -m pip"
+# enforce our chosed pip compile flags
+export COMPILE := BIN + "/pip-compile --allow-unsafe --generate-hashes"
+
+
 # list available commands
 default:
-    @just --list
+    @{{ just_executable() }} --list
+
+
+# clean up temporary files
+clean:
+    rm -rf .venv
 
-# Set up local dev environment
-dev_setup:
+
+# ensure valid virtualenv
+virtualenv:
     #!/usr/bin/env bash
-    set -euo pipefail
-    . scripts/setup_functions
-    dev_setup
+    # allow users to specify python version in .env
+    PYTHON_VERSION=${PYTHON_VERSION:-python3.9}
+    # create venv and upgrade pip
+    test -d $VIRTUAL_ENV || { $PYTHON_VERSION -m venv $VIRTUAL_ENV && $PIP install --upgrade pip; }
+    # ensure we have pip-tools so we can run pip-compile
+    test -e $BIN/pip-compile || $PIP install pip-tools
+
+
+# update requirements.prod.txt if requirement.prod.in has changed
+requirements-prod: virtualenv
+    @test requirements.prod.in -ot requirements.prod.txt || $COMPILE --output-file=requirements.prod.txt requirements.prod.in
```

Comment:
Is this `test` as in the bash function?

Does `@` stop the line from echoing like in bash?

Is there a long version of `-ot` that could use here?

Am I right that the test command is checking something about the .in and .txt (.in is older than .txt?) and then running compile if it fails?

Reply:
1) yes, `/usr/bin/test`
2) yes @ is like in Make
3) no, sadly not, I checked. `ot` is `older than` 
4) yes. I will add a comment to explain.

Reply:
It's basically doing what make does by default with file mtimes. But just dependencies can not be files, so we manually check and exit early.

It's a little clunky, but we generally don't need to compare file mtime's very often, it just happens to be useful when one file is derived from another (i.e. .txt from .in).